### DISCUSSION TITLE
Timer 1 & Timer 3 issues: Timer 1 never triggers | Timer 3 crashes 100% of time

### DISCUSSION THREAD

**Discussion by githubtrey** [micropython/micropython#14235] [ESP32] [answered]
Hi,

Trying to run the python code below on an ESP32-C3 results in a strange output/crash.

```py
from machine import Timer

__version__ = '20240403'
            
def timer0_callback(timer):
    print('t0')

def timer1_callback(timer):
    print('t1')

def timer2_callback(timer):
    print('t2')

def timer3_callback(timer):
    print('t3')


def main():
    timer0 = Timer(0)
    timer0.init(period=1000, mode=Timer.PERIODIC, callback=timer0_callback)
    timer1 = Timer(1)
    timer1.init(period=200, mode=Timer.PERIODIC, callback=timer1_callback)
    timer2 = Timer(2)
    timer2.init(period=50, mode=Timer.PERIODIC, callback=timer2_callback)
#     timer3 = Timer(3)
#     timer3.init(period=300, mode=Timer.PERIODIC, callback=timer3_callback)
   
    while True:
        pass

if __name__ == '__main__':
    main()
```

**Outputs/behaviour:**
- Timer 1: never gets triggered, hence no 't1' output: 
MPY: soft reboot
t2
t2
...
t2
t2
t2
t0
t2
t2
...
- Timer 3: if uncommented : ``` timer3 = Timer(3), timer3.init(period=300, mode=Timer.PERIODIC, callback=timer3_callback) ``` the script immediately crashes 100%. 

**Dump (with Timer 3 uncommented):**

---------------------------------------------------------

MPY: soft reboot
Guru Meditation Error: Core  0 panic'ed (Store access fault). Exception was unhandled.

Core  0 register dump:
MEPC    : 0x40047a74  RA      : 0x420024b2  SP      : 0x3fca86e0  GP      : 0x3fc97e00  
TP      : 0x3fc6e96c  T0      : 0x4005890e  T1      : 0x00000000  T2      : 0xffffffff  
S0/FP   : 0x0000000e  S1      : 0x3fcc0e70  A0      : 0x00000000  A1      : 0x31da05cc  
A2      : 0x00000010  A3      : 0x0001f600  A4      : 0x00000008  A5      : 0x600c2000  
A6      : 0x00010000  A7      : 0x00000000  S2      : 0x00010000  S3      : 0x00000000  
S4      : 0x00000010  S5      : 0x00000020  S6      : 0x3fca8708  S7      : 0x3fcc10f4  
S8      : 0x74737973  S9      : 0x00000020  S10     : 0x00000001  S11     : 0x00000000  
T3      : 0x00000058  T4      : 0xffffffff  T5      : 0xffffffff  T6      : 0xffffffff  
MSTATUS : 0x00001881  MTVEC   : 0x40380001  MCAUSE  : 0x00000007  MTVAL   : 0x31da05cc  
MHARTID : 0x00000000  

Stack memory:
3fca86e0: 0x42006c3a 0x42006c00 0x00000000 0x00000000 0x3fca9b34 0x3fca9b10 0xffc1ffff 0x7fffffff
3fca8700: 0x00000000 0x00000003 0x003e0000 0x00000000 0x00000000 0x00000001 0x00000002 0x00000000
3fca8720: 0x3fca877c 0x3c14187c 0x00000004 0x3c141000 0x3c141000 0x00000185 0x3c1418d0 0x3fcaa3bb
3fca8740: 0x3fffffff 0x3ffffffe 0x40000000 0x3fca87a4 0x80000000 0x00000001 0x3fca9b10 0x42006c00
3fca8760: 0x3fc9b05c 0x00000000 0x00010000 0x00000001 0x3fca9940 0x0000012c 0x000003e8 0x00000006
3fca8780: 0x3fffffff 0x3ffffffe 0x40000000 0x00000003 0x3fca9a08 0x00000001 0x3c141884 0x4200f440
3fca87a0: 0x3fffffff 0x0000001f 0x00000003 0x3fca9a0c 0x000000c0 0x000005a6 0x00002d32 0x3fcaa3d0
3fca87c0: 0x000000c0 0x3fca9a04 0x3fcaa3bc 0x40382fac 0x3c14f000 0x3c156000 0x3fca9a34 0x3fca99f0
3fca87e0: 0x3c14888c 0x3fc9b05c 0x3fca9a30 0x00000000 0x3c149a0c 0x3fca8848 0x00000011 0x3fca9a28
3fca8800: 0x00000020 0x00000000 0x3fca89d8 0x3c149a0c 0x40382994 0x3fca89a0 0x3fc9b0b8 0x00000000
3fca8820: 0x00000000 0x3fca8b58 0x0000000c 0x00000030 0x3fca99f0 0x00000044 0x3c141000 0x3c141000
3fca8840: 0x3c141000 0x3fca87d0 0x00000000 0x00000000 0x00000319 0x00000008 0x00000007 0x4203bbb4
3fca8860: 0x30207970 0x01943f00 0x3f2c0000 0x3fca98e0 0x4039003a 0x4038aa96 0x3fca88f0 0x00000007
3fca8880: 0x3c1456a0 0x00000003 0x3fca9420 0x00000003 0x3fca9420 0x00000004 0x00200940 0x40396a4c
3fca88a0: 0x00000080 0x00000010 0x3fca88b0 0x00200901 0x62696c2f 0x63616d2f 0x656e6968 0x79706d2e
3fca88c0: 0x00797000 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0xffffffff 0x00000080
3fca88e0: 0x4039003a 0x40391804 0x3fca8960 0x3fc97e00 0x3fc6e96c 0xffffffff 0x3c1716b4 0x00000000
3fca8900: 0x00000004 0x00000004 0x00000001 0x00000001 0x00000000 0x0000000c 0x00000004 0x0000000a
3fca8920: 0x3fc9b0bc 0x0000000c 0x00000030 0x0000000c 0x00000000 0x0000b7ce 0x00000001 0x3c141000
3fca8940: 0x3c141000 0x3c141000 0x00000044 0x3fca99f0 0x00000030 0x0000000c 0x3fca8b58 0x00000000
3fca8960: 0x00000000 0x3fc9b0b8 0x3fca89a0 0x4200f390 0x3fffffff 0x3ffffffe 0x3fca8b40 0x3fcaa33b
3fca8980: 0x3fffffff 0x3ffffffe 0x40000000 0x3fcaa3d0 0x000000c0 0x3fca8b54 0x3fcaa33c 0x4038302c
3fca89a0: 0x00000002 0x3fcaa300 0x3fca8b5c 0x3fca8b40 0x3c14888c 0x3fc9b05c 0x3fca8b58 0x00000000
3fca89c0: 0x70656564 0x0000c026 0x00000001 0x3fca8b50 0x0000001f 0x0000001d 0x3fca8bc8 0x000000a7
3fca89e0: 0x40382994 0x3fca8ba0 0x3fc9b0b8 0x00000000 0x00000000 0x00000000 0x00000002 0x00000000
3fca8a00: 0x3fca8b40 0x00000030 0x3c141000 0x3c141000 0x3c141000 0x3fca89a0 0x3fca8b48 0x42041664
3fca8a20: 0x00000000 0x000000a6 0x06ca9940 0x42040000 0x00000000 0x3fca8b20 0x00000000 0x4203f410
3fca8a40: 0x3c141000 0x0000c654 0x00000001 0x3fcaa1c0 0x3c141000 0x3c141000 0x00000001 0x3fcaa1c8
3fca8a60: 0x00000030 0x3fcaa1d4 0x00000032 0x3c141000 0x3c141000 0x3c141000 0x0000011a 0x42049dd8
3fca8a80: 0x00000030 0x3fcaa1d4 0x0000011a 0x4203c114 0x3fca8b20 0x00000002 0x00000001 0x00000000
3fca8aa0: 0x00000002 0xffffffff 0x00000000 0x00000001 0x00000000 0x3fcaa218 0x3fca8b20 0x4203c3b2
3fca8ac0: 0x00000048 0x000000e9 0x3fcaa22c 0x00000009 0x3fca8b20 0x3fcaa258 0x00000008 0x4203c454

ELF file SHA256: 03a427956fd9ca12

Rebooting...
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x403807a8
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd5820,len:0xe8c
load:0x403cc710,len:0x6ec
load:0x403ce710,len:0x2b24
entry 0x403cc710

---------------------------------------------------------

Using the following HW/SW:
- MicroPython v1.22.1 on 2024-01-05; ESP32C3 module with ESP32C3
- SeeedStudio XIAO-ESP32-C3

Any ideas are highly appreciated,

Tom

**Comment by robert-hh** [comment]
Tight busy loops are critical in that they do not allow other tasks to run. Try for the wait loop:

```
while True:
    idle()
```
You have to import idle in at the start of your script as well. Instead of idle, you can as well use a short sleep_ms(), like sleep_ms(1).

**Reply by githubtrey** [reply]
Thank you. 

However, unfortunately it does not change anything at all. I tried idle(), sleep_ms(1), sleep(1.0) or e.g. print('main')... nothing :-(

Any other ideas?

**Reply by robert-hh** [reply]
That's right. I tested that on a Generic ESP32, and there it works. Using soft timers on the ESP32C3 causes a core dump as well. Looks like a Bug.

**Reply by robert-hh** [reply]
Tried that on an S2 and S3, Works well. I used a slightly modified code. On c3, timer0 and timer2 run, timer1 stalls and attempts to start timer3 causes a core dump.
```
from machine import Timer, idle
import time

__version__ = '20240403'

c0 = c1 = c2 = c3 = 0

def timer0_callback(timer):
    global c0
    c0 += 1

def timer1_callback(timer):
    global c1
    c1 += 1

def timer2_callback(timer):
    global c2
    c2 += 1

def timer3_callback(timer):
    global c3
    c3 += 1


def main():
    timer0 = Timer(0, period=1000, mode=Timer.PERIODIC, callback=timer0_callback)
    timer1 = Timer(1, period=200, mode=Timer.PERIODIC, callback=timer1_callback)
    timer2 = Timer(2, period=50, mode=Timer.PERIODIC, callback=timer2_callback)
    # timer3 = Timer(3, period=500, mode=Timer.PERIODIC, callback=timer3_callback)

    try:
        while True:
            print(f"c0: {c0} c1: {c1} c2: {c2} c3: {c3}")
            time.sleep(1)
    except KeyboardInterrupt:
        pass

    timer0.deinit()
    timer1.deinit()
    timer2.deinit()
    # timer3.deinit()

if __name__ == '__main__':
    main()
```

**Reply by robert-hh** [reply, marked as answer]
Reading the C3 manual and looking into the ESP port code reveals:
- the ESP32C3 has just 2 general purpose timers in a single group.
- the generic ESP32 and the S2/S3 devices have 4 timers in 2 groups.
- at the ESP32 port, the timer index is timer_nr *2 + timer_group. So the even numbered timers ar in group 0, the odd ones in group 1.

Thus, on  C3, you can use timer(0) and timer(2). Anything else is odd and should be prevented by the code. Maybe it's better on a C3 to just allow timer 0 and 1 and set the group always to 0.

**Comment by githubtrey** [comment]
Hi Robert,

Thank you for your great help. Seems I made a beginner mistake (assuming ESP32 covers all ESP32 subtypes and IDE not complaining on init of Timers 1 and 3, even aware of C3 variant and not checking the espressif datasheet).

Have a good time

Tom

**Reply by robert-hh** [reply]
I prepared a fix and made a PR with it. It will be timers 0 and 1 when it's eventually merged.

**Comment by dmoibm** [comment]
The ESP32 C6 has the same behavior as the C3. Only Timer(0) and Timer(2) are allowed. 

Things get fun with the use of Timer(4), which overwrites Timer(0) and so on.

**Comment by robert-hh** [comment]
@dmoibm That behavior was fixed with commit 136058496 , PR #16439 in December. Using a recent preview only Timers 0 and 1 are allowed for ESP32C6.
